### ISSUE TITLE
Add a parser for time-dependent boundary conditions

### ISSUE THREAD

**Issue by charlieowen3065** [UT-Computational-NE/FlowForge#56]
## Description

For boundary conditions in transient systems, it is common to have the boundary values also time-dependent (a heating pipe, an increasing flow rate, etc.). To do this, FlowForge will need a method to (1) take in a string-type equation and (2) parse this into an evaluate-able function.

Then, the boundary conditions should be updated with the functionality to take in these time-dependent equations and pass them through as objects (passing them as object where previously they would have just been constants).

## Acceptance Criteria

- [x] Create parser object
- [x] Modify the boundary condition object to accept this function

## Definition of Done

- [x] All acceptance criteria are met
- [x] Task has been reviewed and approved by a supervisor/team lead
- [x] Any necessary documentation has been completed
- [x] Task outcomes have been tested and verified

**Comment by charlieowen3065** [comment]
For the input file, the only difference for boundary conditions will be that, instead of just the input `value`, a `function` will also be able to be input.

In practice, this new input file should look similar to
```
time_dependent_temperature : {
    "boundary_type" : "DirichletBC",
    "surface": "inlet", 
    "variable" : "temperature",
    "function" : '700 + 5*t'
}
```
where the input file now can take `function` or `value`.

**Comment by charlieowen3065** [comment]
The object that takes in these string equations (such as `5*t + 10`) will have 4 hard-coded variables:

1. `x` : the x-coordinate of interest
2. `y` : the y-coordinate of interest
3. `z` : the z-coordinate of interest
4. `t` : an input time

For now, the object can also take in a list of variable names (such as `['temperature', 'pressure']`) which would represent coupled variables to the system. For now, it will stay like that as a reminder that it is (1) possible and (2) should be later implemented (for uses such as user defined density, thermal conductivity, etc which would be functions of temperature, pressure, etc.).

**Comment by charlieowen3065** [comment]
The `EquationParser` object will have the public method:
```
evaluate(x=None, y=none, z=None, t=None, **kwargs)
```
where `**kwargs` is a stand-in for the still needed coupled variables.

For now, this method checks if each variable is not `None`, and substitutes the input value in the expression. This is done for each variable until fully evaluated, and the function outputs a float.

TODO: Get rid of 'if' statements to avoid checking ifs during runtime.

**Comment by cgentry7** [comment]
In Python, you don't need to check if a variable is `None` or not, you just need to go ahead and evaluate the function string, and let the evaluator throw an error if the variable is improperly assigned.  I would imagine `smpy` should be able to handle this ok.

**Comment by charlieowen3065** [comment]
@cgentry7 

In sympy, attempting to substitute in a variable of type `None` will throw an error to the system. To get the `evaluate` function working without if-statements would either require initializing the variables as `0` instead of `None` (which would avoid this issue) or adding functionality to parse through which variables were input and which were valid. 

I decided on the latter method as I would like some way of checking that the inputs into `evaluate` were valid, and setting them to zero would allow for easier errors to be made. To do this parsing, I have added a method to extract valid variable names during initialization, and added a method that only gives these valid variables to the `evaluate` function.

In this manner, any non-defined variables will be discarded, and a forgotten variable will be kept as `None` and will throw an error to the user, which is desired.
